packages/vm/src/state/transientStorage.ts

```diff
@@ -0,0 +1,131 @@
+import { Address } from 'ethereumjs-util'
+
+export type TStorage = Map<string, Map<string, Buffer>>
+
+export interface TransientStorageModification {
+  addr: Address
+  key: Buffer
+  prevValue: Buffer
+}
+
+export type Changeset = TransientStorageModification[]
+
+export interface TransientStorageOptions {
+  storage?: TStorage
+  changesets?: Changeset[]
+}
+
+export default class TransientStorage {
+  _storage: TStorage
+  _changesets: Changeset[]
+
+  constructor(opts: TransientStorageOptions = {}) {
+    this._storage = opts.storage ?? new Map()
+    this._changesets = opts.changesets ?? [[]]
+  }
+
+  private addModification(modification: TransientStorageModification) {
+    if (this._changesets.length === 0) {
+      throw new Error('no changeset initialized')
+    }
+    const latest = this._changesets[this._changesets.length - 1]
+    latest.push(modification)
+  }
+
+  public get(addr: Address, key: Buffer): Buffer {
+    const map = this._storage.get(addr.toString())
+    if (!map) {
+      return Buffer.alloc(32, 0x00)
+    }
+    const value = map.get(key.toString('hex'))
+    if (!value) {
+      return Buffer.alloc(32, 0x00)
+    }
+    return value
+  }
+
+  public put(addr: Address, key: Buffer, value: Buffer) {
+    if (key.length !== 32) {
+      throw new Error('Transient storage key must be 32 bytes long')
+    }
+
+    if (value.length > 32) {
+      throw new Error('Transient storage value cannot be longer than 32 bytes')
+    }
+
+    if (!this._storage.has(addr.toString())) {
+      this._storage.set(addr.toString(), new Map())
+    }
+    const map = this._storage.get(addr.toString())!
+
+    const str = key.toString('hex')
+    const prevValue = map.get(str) ?? Buffer.alloc(32, 0x00)
```

Comment:
Nit: if you use `Buffer.alloc` with only 1 argument then it defaults to fill it with zeros.

Reply:
Will remove the second arg, was being explicit here